tiledb/ml/readers/_batch_utils.py

```diff
@@ -265,43 +189,151 @@ def batch_factory(
+                futures.wait(
+                    (
+                        executor.submit(x_gen.read_buffer, batch.x_read_slice),
+                        executor.submit(y_gen.read_buffer, batch.y_read_slice),
+                    )
                 )
+            elif batch.x_read_slice:
+                x_gen.read_buffer(batch.x_read_slice)
+            elif batch.y_read_slice:
+                y_gen.read_buffer(batch.y_read_slice)
+
+            if batch_shuffle and batch.shuffling:
+                row_idxs = np.arange(batch.shuffling.size)
+                np.random.shuffle(row_idxs)
+                x_gen.shuffle_buffer(batch.shuffling.x_buffer_slice, row_idxs)
+                y_gen.shuffle_buffer(batch.shuffling.y_buffer_slice, row_idxs)
+
+            x_tensors = x_gen.iter_tensors(batch.x_buffer_slice)
+            y_tensors = y_gen.iter_tensors(batch.y_buffer_slice)
+            yield (*x_tensors, *y_tensors)
+
+
+@dataclass(frozen=True, repr=False)
+class Shuffling:
+    size: int
+    x_buffer_slice: slice
+    y_buffer_slice: slice
+
+    def __repr__(self) -> str:
+        x, y = self.x_buffer_slice, self.y_buffer_slice
+        return f"Shuffling(x[{x.start}:{x.stop}], y[{y.start}:{y.stop}])"
+
+
+@dataclass(frozen=True, repr=False)
+class Batch:
+    x_read_slice: Optional[slice]
+    y_read_slice: Optional[slice]
+    shuffling: Optional[Shuffling]
+    x_buffer_slice: slice
+    y_buffer_slice: slice
+
+    def __repr__(self) -> str:
+        x, y = self.x_buffer_slice, self.y_buffer_slice
+        s = f"Batch(x[{x.start}:{x.stop}], y[{y.start}:{y.stop}]"
+        if self.x_read_slice:
+            s += f", x_read[{self.x_read_slice.start}:{self.x_read_slice.stop}]"
+        if self.y_read_slice:
+            s += f", y_read[{self.y_read_slice.start}:{self.y_read_slice.stop}]"
+        if self.shuffling:
+            s += f", {self.shuffling}"
+        return s + ")"
+
+
+def iter_batches(
+    batch_size: int,
+    x_buffer_size: int,
+    y_buffer_size: int,
+    start_offset: int,
+    stop_offset: int,
+) -> Iterator[Batch]:
+    """
+    Generate `Batch` instances describing each batch.
+
+    Each yielded `Batch` instance describes:
+    - The slice to read from the x array into x buffer (if the current x buffer is consumed).
+    - The slice to read from the y array into y buffer (if the current y buffer is consumed).
+    - The shuffling to apply to x and y buffers (if there is new x and/or y buffer).
+      - How many buffer rows to shuffle.
+      - The slice of the x buffer to shuffle.
+      - The slice of the y buffer to shuffle.
+    - The batch slice to read from the x buffer.
+    - The batch slice to read from the y buffer.
+
+    :param batch_size: (Max) size of each batch.
+    :param x_buffer_size: (Max) size of the x buffer.
+    :param y_buffer_size: (Max) size of the y buffer.
+    :param start_offset: Start row offset.
+    :param stop_offset: Stop row offset.
+    """
+    assert (
```

Comment:
Since we care that this is raised in any scenario, would not be better to raise Exception instead? Except for that LGTM.

Reply:
Raising an exception would be needed If this was a public function to be called by the library user. In this case it's an internal function that we call by passing appropriate buffer sizes, hence the assertion.